server/src/test/java/io/crate/integrationtests/GeoShapeIntegrationTest.java

```diff
@@ -251,4 +250,32 @@ public void test_geo_shape_can_be_used_as_generated_column() throws Exception {
             "{coordinates=[10, 20], type=point}"
         );
     }
+
+    @Test
+    public void test_polygon_can_be_used_as_generated_column() {
+        // Dynamic nested array is currently not supported, declaring coordinates array upfront.
+        execute("""
+            create table tbl
+            (
+                o object as (
+                    o1 object as (coordinates ARRAY(ARRAY(ARRAY(DOUBLE))), type string),
+                    geo as o['o1']::geo_shape
+               )
+            )
+            """
+        );
+        List<List<List<Double>>> coords = List.of(
+            List.of(List.of(0.0, 0.0), List.of(1.0, 0.0), List.of(1.0, 1.0), List.of(0.0, 0.0))
+        );
+        execute(
+            "insert into tbl (o) values (?)",
+            new Object[] {
+                Map.of("o1", Map.of("coordinates", coords, "type", "Polygon"))
+            }
+        );
+        execute("refresh table tbl");
+        assertThat(execute("select o['geo'] from tbl")).hasRows(
+            "{coordinates=[[[0.0, 0.0], [1.0, 0.0], [1.0, 1.0], [0.0, 0.0]]], type=Polygon}"
+        );
+    }
```

Comment:
I decided to add a test because we already had a case of fixing one part of generated column aspect and missing some scenario due to refactoring or something else. See https://github.com/crate/crate/pull/16378#issuecomment-2247377927